# STEPS 1-3

## IMPORTING LIBRARIES

In [2]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## PATH

In [3]:
path = r'C:\Users\gryzu\12.03.2023 Instacart Basket Analysis'

In [4]:
path

'C:\\Users\\gryzu\\12.03.2023 Instacart Basket Analysis'

## IMPORTING CUSTOMER DATASET

In [5]:
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [6]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# STEP 4

In [7]:
# renaming several columns
df_customers.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [8]:
df_customers.rename(columns = {'surname' : 'last_name'}, inplace = True)

In [9]:
df_customers.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [10]:
df_customers.rename(columns = {'STATE' : 'state'}, inplace = True)

In [11]:
df_customers.rename(columns = {'Age' : 'age'}, inplace = True)

In [12]:
df_customers.rename(columns = {'n_dependants' : 'number_of_dependants'}, inplace = True)

In [13]:
df_customers.rename(columns = {'fam_status' : 'family_status'}, inplace = True)

In [14]:
# Illustrating an updated table
df_customers.head()

user_id first_name    Surnam  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants family_status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

# STEP 5

## CHECKING MISSING VALUES

In [15]:
df_customers.isnull().sum()

user_id                     0
first_name              11259
Surnam                      0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
family_status               0
income                      0
dtype: int64

In [16]:
df_nan = df_customers[df_customers['first_name'].isnull() == True]

In [17]:
df_nan

user_id first_name    Surnam  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        number_of_dependants family_status  income  
53                         2       married   41709  
73                         0        single   82518  
82                         3       married  117099  
99                         1       married  155673  
105                        3       married  151819  
...                      ...           ...     ...  
206038                     3       married   87783  
206044                     2       married  108488  
206090                     1       married   45275  
206162                     0        single   56325  
206171                     2       married   59222  

[11259 rows x 10 columns]

We have 11259 unknown values, where first names are missing. We will keep the data as it is, since removing it will negatively affect our final results.

## CHECKING FOR DUPLICATES 

In [18]:
df_dups = df_customers[df_customers.duplicated()]

In [19]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, Surnam, gender, state, age, date_joined, number_of_dependants, family_status, income]
Index: []

NO DUPLICATES HAVE BEEN FOUND

## DESCRIPTIVE STATISTICS CHECK

In [20]:
df_customers.describe()

user_id            age  number_of_dependants         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

Statistical indicators look good. No deviations are to be seen.

## CHECKING DATA TYPES

In [21]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   first_name            194950 non-null  object
 2   Surnam                206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   number_of_dependants  206209 non-null  int64 
 8   family_status         206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [22]:
# changing date_joined from object to timeline 
df_customers['date_joined']=pd.to_datetime(df_customers['date_joined'])

In [23]:
df_customers.dtypes

user_id                          int64
first_name                      object
Surnam                          object
gender                          object
state                           object
age                              int64
date_joined             datetime64[ns]
number_of_dependants             int64
family_status                   object
income                           int64
dtype: object

# STEPS 6-9

In [25]:
# Importing the ords_prods_merge dataset 
path_to_pickle = os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_grouped.pkl')

In [26]:
df_ords_prods_merge = pd.read_pickle(path_to_pickle)

In [27]:
# Checking Table
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge     busiest_day    busiest_days  \
0         Soda  ...    both  Regularly busy  Regularly busy   
1         Soda  ...    both  Regularly busy    Slowest days   
2         Soda  ...    both  Regularly busy    Slowest days   
3         Soda  ...    both      Least busy    Slowest days   
4         Soda  ...    both      Least busy    Slowest days   

  busiest_period_of_day max_order  loyalty_flag mean_product_price  \
0        Average orders        10  New customer           6.367797   
1        Average orders        10  New customer           6.367797   
2           Most orders        10  New customer           6.367797   
3        Average orders        10  New customer           6.367797   
4           Most orders        10  New customer           6.367797   

   spending_flag median_days_between_orders        order_frequency  
0    Low spender                       20.5  Non-frequent customer  
1    Low spender                       20.5  Non-frequent customer  
2    Low spender                       20.5  Non-frequent customer  
3    Low spender                       20.5  Non-frequent customer  
4    Low spender                       20.5  Non-frequent customer  

[5 rows x 23 columns]

In [29]:
#Removing 'merge' column in order to make a merge
df_ords_prods_merge1 = df_ords_prods_merge.drop(['_merge'], axis=1)

In [30]:
df_ords_prods_merge1.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices     busiest_day    busiest_days  \
0         Soda  ...     9.0  Regularly busy  Regularly busy   
1         Soda  ...     9.0  Regularly busy    Slowest days   
2         Soda  ...     9.0  Regularly busy    Slowest days   
3         Soda  ...     9.0      Least busy    Slowest days   
4         Soda  ...     9.0      Least busy    Slowest days   

  busiest_period_of_day max_order  loyalty_flag  mean_product_price  \
0        Average orders        10  New customer            6.367797   
1        Average orders        10  New customer            6.367797   
2           Most orders        10  New customer            6.367797   
3        Average orders        10  New customer            6.367797   
4           Most orders        10  New customer            6.367797   

  spending_flag  median_days_between_orders        order_frequency  
0   Low spender                        20.5  Non-frequent customer  
1   Low spender                        20.5  Non-frequent customer  
2   Low spender                        20.5  Non-frequent customer  
3   Low spender                        20.5  Non-frequent customer  
4   Low spender                        20.5  Non-frequent customer  

[5 rows x 22 columns]

In [39]:
#checking the datatypes
df_ords_prods_merge1.dtypes

order_id                        int64
user_id                         int64
order_number                    int64
orders_day_of_week              int64
order_hour_of_day               int64
days_since_prior_order        float64
product_id                      int64
add_to_cart_order               int64
reordered                       int64
product_name                   object
aisle_id                        int64
department_id                   int64
prices                        float64
busiest_day                    object
busiest_days                   object
busiest_period_of_day          object
max_order                       int64
loyalty_flag                   object
mean_product_price            float64
spending_flag                  object
median_days_between_orders    float64
order_frequency                object
dtype: object

user_id will serve as a great key to combine both datasets: df_customers and df_ords_prods_merge1

## MERGING TWO DATASETS 

In [33]:
df_merged = pd.merge(df_customers, df_ords_prods_merge1, on='user_id', indicator=True)

In [38]:
df_merged.head()

user_id first_name    Surnam  gender     state  age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
1    26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
2    26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
3    26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
4    26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   

   number_of_dependants family_status  income  ...     busiest_day  \
0                     3       married  165665  ...  Regularly busy   
1                     3       married  165665  ...  Regularly busy   
2                     3       married  165665  ...  Regularly busy   
3                     3       married  165665  ...  Regularly busy   
4                     3       married  165665  ...  Regularly busy   

     busiest_days  busiest_period_of_day  max_order  loyalty_flag  \
0    Busiest days            Most orders          8  New customer   
1  Regularly busy            Most orders          8  New customer   
2    Busiest days            Most orders          8  New customer   
3  Regularly busy            Most orders          8  New customer   
4    Slowest days            Most orders          8  New customer   

   mean_product_price  spending_flag  median_days_between_orders  \
0            7.988889    Low spender                        19.0   
1            7.988889    Low spender                        19.0   
2            7.988889    Low spender                        19.0   
3            7.988889    Low spender                        19.0   
4            7.988889    Low spender                        19.0   

    order_frequency  _merge  
0  Regular customer    both  
1  Regular customer    both  
2  Regular customer    both  
3  Regular customer    both  
4  Regular customer    both  

[5 rows x 32 columns]

## EXPORTING DATA

In [40]:
df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'customer_orders_products.pkl'))